types/d3-array/index.d.ts

```diff
@@ -311,29 +314,64 @@ export function zip<T>(...arrays: Array<ArrayLike<T>>): T[][];
 // Histogram
 // --------------------------------------------------------------------------------------
 
-export interface Bin<Datum, Value extends number | Date> extends Array<Datum> {
-    x0: Value;
-    x1: Value;
+export interface Bin<Datum, Value extends number | Date | undefined> extends Array<Datum> {
+    x0: Value | undefined;
```

Comment:
`Value` is the value accessor return type. If someone use a value accessor which always returns `number` and also may pass an empty list to `histogram(.)` at some point, we will force him to use `number | undefined` instead as `Value` generic.

Also note that all over in d3-array, we have functions returning `undefined` on empty list, and which therefore need a type assertion for most common use. I cannot tell you how many times I wrote: `d3.max(...)!` or `d3.min(...)!`. So I find this consistant.

Of course in general and for most users, it would be simpler as you propose...

Reply:
As you point out the scenario is slightly different from the one in **d3-geo**, i.e. the empty list argument you make w.r.t. `histogram(...)`, so let's leave it as is here.

I know, what you are saying about `!`-assertions in other uses of **d3-array**, I just prefer to minimize the need for it, whenever, there is a well-defined declarative approach.

More broadly, I find the assertions can easily get out of hand and can mask mistakes, in particular when re-factoring upstream code which then violates the assumptions that once justified the assertion. :smile: